<a href="https://colab.research.google.com/github/ai-nlp-gl/ticket-assignment/blob/Adarsh/Adarsh_nlp_GIT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#First let us mount the google drive

In [0]:
from google.colab import drive
drive.mount('/content/drive/')

Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).


checking for current directory and set current directory path if necessary

In [0]:
import os
os.getcwd()

'/content/drive/My Drive/Capstone_Project/Ticket_Assignment'

In [0]:
!ls 

 input_data.xlsx
'Project - NLP - Problem Statement - Automatic Ticket Assignment.pdf'


In [0]:
os.chdir('/content/drive/My Drive/Capstone_Project/Ticket_Assignment')

In [0]:
!ls

 input_data.xlsx
'Project - NLP - Problem Statement - Automatic Ticket Assignment.pdf'


In [0]:
os.getcwd()

'/content/drive/My Drive/Capstone_Project/Ticket_Assignment'

Let us import basic packages

In [0]:
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
%matplotlib inline

Let us import the input data to be analyzed

In [0]:
df = pd.read_excel('input_data.xlsx')

**1.Exploratory Data Analysis**



In [0]:
df.head(10)

Short description  ... Assignment group
0                                        login issue  ...            GRP_0
1                                            outlook  ...            GRP_0
2                                 cant log in to vpn  ...            GRP_0
3                      unable to access hr_tool page  ...            GRP_0
4                                       skype error   ...            GRP_0
5     unable to log in to engineering tool and skype  ...            GRP_0
6  event: critical:HostName_221.company.com the v...  ...            GRP_1
7  ticket_no1550391- employment status - new non-...  ...            GRP_0
8               unable to disable add ins on outlook  ...            GRP_0
9                    ticket update on inplant_874773  ...            GRP_0

[10 rows x 4 columns]

In [0]:
df.shape

(8500, 4)

In [0]:
df.describe().transpose()




count unique                top  freq
Short description  8492   7481     password reset    38
Description        8499   7814                the    56
Caller             8500   2950  bpctwhsn kzqsbmtp   810
Assignment group   8500     74              GRP_0  3976

In [0]:
df.isnull().sum()

Short description    8
Description          1
Caller               0
Assignment group     0
dtype: int64

**1.Exploratory Data Analysis **
   **--> Null value Handling **

Clearly there are null values that needs to be handled

Let us locate where are the null values present in the respective columns

In [0]:
print(df[df['Description'].isnull()])

             Short description Description             Caller Assignment group
4395  i am locked out of skype         NaN  viyglzfo ajtfzpkb            GRP_0


In [0]:
print(df[df['Short description'].isnull()])

     Short description  ... Assignment group
2604               NaN  ...           GRP_34
3383               NaN  ...            GRP_0
3906               NaN  ...            GRP_0
3910               NaN  ...            GRP_0
3915               NaN  ...            GRP_0
3921               NaN  ...            GRP_0
3924               NaN  ...            GRP_0
4341               NaN  ...            GRP_0

[8 rows x 4 columns]


Let us observe some other details before handling null values

In [0]:
df.dtypes

Short description    object
Description          object
Caller               object
Assignment group     object
dtype: object

In [0]:
df['Caller'].value_counts()

bpctwhsn kzqsbmtp    810
ZkBogxib QsEJzdZO    151
fumkcsji sarmtlhy    134
rbozivdq gmlhrtvp     87
rkupnshb gsmzfojw     71
                    ... 
wgmnjxft zeyhasdj      1
tnhzdapi juqcbdzl      1
wohzmlib fxwjhapo      1
mizpywld dnrubpis      1
gjisfonb odwfhmze      1
Name: Caller, Length: 2950, dtype: int64

In [0]:
df['Assignment group'].value_counts()

GRP_0     3976
GRP_8      661
GRP_24     289
GRP_12     257
GRP_9      252
          ... 
GRP_70       1
GRP_35       1
GRP_73       1
GRP_67       1
GRP_61       1
Name: Assignment group, Length: 74, dtype: int64

In [0]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8500 entries, 0 to 8499
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Short description  8492 non-null   object
 1   Description        8499 non-null   object
 2   Caller             8500 non-null   object
 3   Assignment group   8500 non-null   object
dtypes: object(4)
memory usage: 265.8+ KB


In [0]:
df.iloc[2604]

Short description                                                  NaN
Description          \n\nreceived from: ohdrnswl.rezuibdt@gmail.com...
Caller                                               ohdrnswl rezuibdt
Assignment group                                                GRP_34
Name: 2604, dtype: object

In [0]:
df.iloc[3383]

Short description                                                  NaN
Description          \n-connected to the user system using teamview...
Caller                                               qftpazns fxpnytmk
Assignment group                                                 GRP_0
Name: 3383, dtype: object

In [0]:
df.iloc[3906]

Short description                                                  NaN
Description          -user unable  tologin to vpn.\n-connected to t...
Caller                                               awpcmsey ctdiuqwe
Assignment group                                                 GRP_0
Name: 3906, dtype: object

In [0]:
df.iloc[3910]

Short description                                                  NaN
Description          -user unable  tologin to vpn.\n-connected to t...
Caller                                               rhwsmefo tvphyura
Assignment group                                                 GRP_0
Name: 3910, dtype: object

In [0]:
df.iloc[3915]

Short description                                                  NaN
Description          -user unable  tologin to vpn.\n-connected to t...
Caller                                               hxripljo efzounig
Assignment group                                                 GRP_0
Name: 3915, dtype: object

In [0]:
df.iloc[3921]

Short description                                                  NaN
Description          -user unable  tologin to vpn.\n-connected to t...
Caller                                               cziadygo veiosxby
Assignment group                                                 GRP_0
Name: 3921, dtype: object

In [0]:
df.iloc[3924]

Short description                                                  NaN
Description          name:wvqgbdhm fwchqjor\nlanguage:\nbrowser:mic...
Caller                                               wvqgbdhm fwchqjor
Assignment group                                                 GRP_0
Name: 3924, dtype: object

In [0]:
df.iloc[4341]

Short description                                                  NaN
Description          \n\nreceived from: eqmuniov.ehxkcbgj@gmail.com...
Caller                                               eqmuniov ehxkcbgj
Assignment group                                                 GRP_0
Name: 4341, dtype: object

In [0]:
df.iloc[4395]

Short description    i am locked out of skype
Description                               NaN
Caller                      viyglzfo ajtfzpkb
Assignment group                        GRP_0
Name: 4395, dtype: object

In [0]:
df['Short description'].str.find('i am locked out of skype').value_counts

<bound method IndexOpsMixin.value_counts of 0      -1.0
1      -1.0
2      -1.0
3      -1.0
4      -1.0
       ... 
8495   -1.0
8496   -1.0
8497   -1.0
8498   -1.0
8499   -1.0
Name: Short description, Length: 8500, dtype: float64>

In [0]:
df.iloc[4355:5000].head(50)

Short description  ... Assignment group
4355                                         erp system  ...            GRP_0
4356                                                erp  ...            GRP_0
4357                                            re; erp  ...            GRP_0
4358  an terminal 12 bei iso-u können im EU_tool ke...  ...           GRP_42
4359                                        bobj error   ...            GRP_9
4360  job bkwin_tax_interface_qa_daily failed in job...  ...            GRP_8
4361                               account locked in ad  ...            GRP_0
4362  delete the charm 8111117934,  since it is no l...  ...           GRP_38
4363  job pp_EU_tool_netch_ap2 failed in job_schedul...  ...            GRP_8
4364  job bkwin_HostName_770_inc failed in job_sched...  ...            GRP_5
4365  unable to load engineering_tool on my new 2-1 ...  ...            GRP_0
4366           vpn and vpn will not allow access to erp  ...            GRP_0
4367                         engineering_tool phone app  ...           GRP_51
4368  can you please remove quyhn mcgudftigre from u...  ...           GRP_26
4369  since last security update i have been unable ...  ...            GRP_0
4370  application response time (other network resou...  ...            GRP_4
4371                                  outlook/crm error  ...            GRP_0
4372  all contacts loaded for imts -existing account...  ...           GRP_40
4373  job SID_37hotf failed in job_scheduler at: 09/...  ...            GRP_5
4374                                        login issue  ...            GRP_0
4375  office 2013 for qohfjpna exphkims shows activa...  ...           GRP_26
4376                                             remove  ...            GRP_0
4377               error login on to the SID_34 system.  ...            GRP_0
4378  need to change password and get it synced on s...  ...            GRP_0
4379  outlook not allowing to open emails in the inbox.  ...            GRP_0
4380  external monitors not detecting with dell 2in ...  ...            GRP_0
4381                          erp SID_34 account locked  ...            GRP_0
4382                printer problem / issue information  ...            GRP_3
4383                          erp SID_39 password reset  ...            GRP_0
4384                               printer driver error  ...            GRP_0
4385  pc set up for new employee -mghlisha.baranwfhr...  ...           GRP_19
4386                                         wifi issue  ...           GRP_19
4387                                        login issue  ...            GRP_0
4388                                         spam email  ...            GRP_0
4389                        pc set up for new employee   ...           GRP_19
4390                                        pdf printer  ...           GRP_19
4391                                  change the old pc  ...           GRP_19
4392                           frequent account lockout  ...            GRP_2
4393  mii uacyltoe hxgaycze from lacw8515941  : dura...  ...           GRP_14
4394           approved expense report date range error  ...           GRP_10
4395                           i am locked out of skype  ...            GRP_0
4396  unable to host a skype meeting from a conferen...  ...            GRP_0
4397                        vip2: printer driver update  ...            GRP_0
4398  erp user blocked, please unlock immediately & ...  ...            GRP_0
4399       password reset from password_management_tool  ...            GRP_0
4400                                     password reset  ...            GRP_0
4401                              problem with printing  ...            GRP_0
4402            pricing error on order number 35022202.  ...           GRP_13
4403                          erp SID_39 password reset  ...            GRP_0
4404  reset the password for ahydmrbu fjymgtvo on ot...  ...            GRP_7

[50 rows x 4 columns]

After going through the descriptions and short descriptions ,we can observer that most issues of skype are of grp 0.So we can just fill the same info in the empty description of row 4395.

Description info to be filled as 'i am locked out of skype .Caller : viyglzfo ajtfzpkb

In [0]:

df['Description']=df['Description'].mask(pd.isnull,'i am locked out of skype .Caller : viyglzfo ajtfzpkb')
df.iloc[4395]



Short description                             i am locked out of skype
Description          i am locked out of skype .Caller : viyglzfo aj...
Caller                                               viyglzfo ajtfzpkb
Assignment group                                                 GRP_0
Name: 4395, dtype: object

In [0]:
df['Description'].isnull().sum()

0

Now we need to deal with null values in Short description

In [0]:
df.iloc[3922]

Short description                                     vpn login issue.
Description          vpn login issue.\n-user unable  tologin to vpn...
Caller                                               suyighpt wifujdvq
Assignment group                                                 GRP_0
Name: 3922, dtype: object

For most of the null values regarding vpn connectivity ,we can use the same short description shown above as the functionality and description mentioned are similar

In [0]:
print(df[df['Short description'].isnull()])

     Short description  ... Assignment group
2604               NaN  ...           GRP_34
3383               NaN  ...            GRP_0
3906               NaN  ...            GRP_0
3910               NaN  ...            GRP_0
3915               NaN  ...            GRP_0
3921               NaN  ...            GRP_0
3924               NaN  ...            GRP_0
4341               NaN  ...            GRP_0

[8 rows x 4 columns]


In [0]:
df.iloc[3383]

Short description                                                  NaN
Description          \n-connected to the user system using teamview...
Caller                                               qftpazns fxpnytmk
Assignment group                                                 GRP_0
Name: 3383, dtype: object

In [0]:
df.drop([3383,3924,4341,2604],inplace=True)

In [0]:
print(df[df['Short description'].isnull()])

     Short description  ... Assignment group
3906               NaN  ...            GRP_0
3910               NaN  ...            GRP_0
3915               NaN  ...            GRP_0
3921               NaN  ...            GRP_0

[4 rows x 4 columns]


In [0]:
df['Short description'].fillna("vpn login issue.",inplace = True)

In [0]:
df['Short description'].isnull().sum()

0

In [0]:
df.isnull().sum()

Short description    0
Description          0
Caller               0
Assignment group     0
dtype: int64

Null values has been handled.some rows has been dropped and some rows has been filled with description based on similarities with other rows done on comparison

In [0]:
df.shape

(8496, 4)

We have dropped four rows to handle null values,so total no of rows is 8496

In [0]:
df['Assignment group'].value_counts()

GRP_0     3973
GRP_8      661
GRP_24     289
GRP_12     257
GRP_9      252
          ... 
GRP_70       1
GRP_35       1
GRP_73       1
GRP_67       1
GRP_61       1
Name: Assignment group, Length: 74, dtype: int64

In [0]:
df['Caller'].value_counts()

bpctwhsn kzqsbmtp    810
ZkBogxib QsEJzdZO    151
fumkcsji sarmtlhy    134
rbozivdq gmlhrtvp     87
rkupnshb gsmzfojw     71
                    ... 
vrmpysoz qkiucpdx      1
wgmnjxft zeyhasdj      1
tnhzdapi juqcbdzl      1
wohzmlib fxwjhapo      1
gjisfonb odwfhmze      1
Name: Caller, Length: 2949, dtype: int64

In [0]:
df.describe()

Short description Description             Caller Assignment group
count               8496        8496               8496             8496
unique              7481        7811               2949               74
top       password reset         the  bpctwhsn kzqsbmtp            GRP_0
freq                  38          56                810             3973

In [0]:

df.describe().transpose()

count unique                top  freq
Short description  8496   7481     password reset    38
Description        8496   7811                the    56
Caller             8496   2949  bpctwhsn kzqsbmtp   810
Assignment group   8496     74              GRP_0  3973

In [0]:
df.groupby('Assignment group').describe().transpose()

Assignment group                           GRP_0  ...                                              GRP_9
Short description count                     3973  ...                                                252
                  unique                    3201  ...                                                237
                  top             password reset  ...             abended job in job_scheduler: Job_1148
                  freq                        38  ...                                                  5
Description       count                     3973  ...                                                252
                  unique                    3406  ...                                                252
                  top     windows password reset  ...  received from: monitoring_tool@company.com\n\n...
                  freq                        29  ...                                                  1
Caller            count                     3973  ...                                                252
                  unique                    2039  ...                                                 57
                  top          fumkcsji sarmtlhy  ...                                  bpctwhsn kzqsbmtp
                  freq                       132  ...                                                153

[12 rows x 74 columns]

**EDA - Word Cloud**

In [0]:
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
nltk.download('popular')

[nltk_data] Downloading collection 'popular'
[nltk_data]    | 
[nltk_data]    | Downloading package cmudict to /root/nltk_data...
[nltk_data]    |   Package cmudict is already up-to-date!
[nltk_data]    | Downloading package gazetteers to /root/nltk_data...
[nltk_data]    |   Package gazetteers is already up-to-date!
[nltk_data]    | Downloading package genesis to /root/nltk_data...
[nltk_data]    |   Package genesis is already up-to-date!
[nltk_data]    | Downloading package gutenberg to /root/nltk_data...
[nltk_data]    |   Package gutenberg is already up-to-date!
[nltk_data]    | Downloading package inaugural to /root/nltk_data...
[nltk_data]    |   Package inaugural is already up-to-date!
[nltk_data]    | Downloading package movie_reviews to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Package movie_reviews is already up-to-date!
[nltk_data]    | Downloading package names to /root/nltk_data...
[nltk_data]    |   Package names is already up-to-date!
[nltk_data]    | Do

True

In [0]:
stop_words = set(stopwords.words('english'))

In [0]:
stop_words.remove('not')

In [0]:
df1 = df

In [0]:
#df1.drop('Assignment group',axis = 1,inplace = True)

In [0]:
#df1.add('Caller',axis = 1,inplace = True)
  

In [0]:
df1['Description'] = df1['Description'].apply(lambda x: " ".join(x for x in x.split() if x not in stop_words))


In [0]:
df1['Short description'] = df1['Short description'].apply(lambda x: " ".join(x for x in x.split() if x not in stop_words))

In [0]:
df1['Caller'] = df1['Caller'].apply(lambda x: x.split())

In [0]:
df1['Description'] = df1['Description'].apply(lambda x: nltk.RegexpTokenizer(r"\w+").tokenize(x))

In [0]:
df1['Short description'] = df1['Short description'].apply(lambda x: nltk.RegexpTokenizer(r"\w+").tokenize(x))

In [0]:
df1

Short description  ... Assignment group
0                                        [login, issue]  ...            GRP_0
1                                             [outlook]  ...            GRP_0
2                                      [cant, log, vpn]  ...            GRP_0
3                       [unable, access, hr_tool, page]  ...            GRP_0
4                                        [skype, error]  ...            GRP_0
...                                                 ...  ...              ...
8495                    [emails, not, coming, zz, mail]  ...           GRP_29
8496                        [telephony_software, issue]  ...            GRP_0
8497  [vip2, windows, password, reset, tifpdchb, ped...  ...            GRP_0
8498                [machine, nÃ, o, estÃ, funcionando]  ...           GRP_62
8499  [mehreren, pc, s, lassen, sich, verschiedene, ...  ...           GRP_49

[8496 rows x 4 columns]

In [0]:
contractions = { 
"ain't": "am not / are not / is not / has not / have not",
"aint": "am not / are not / is not / has not / have not",
"aren't": "are not / am not",
"arent": "are not / am not",
"can't": "cannot",
"cant": "cannot",
"can't've": "cannot have",
"cant've": "cannot have",
"cantve": "cannot have",
"'cause": "because",
"could've": "could have",
"couldve": "could have",
"couldn't": "could not",
"couldnt": "could not",
"couldn't've": "could not have",
"didn't": "did not",
"didnt": "did not",
"doesn't": "does not",
"doesnt": "does not",
"don't": "do not",
"dont": "do not",
"hadn't": "had not",
"hadnt": "had not",
"hadn't've": "had not have",
"hasn't": "has not",
"hasnt": "has not",
"haven't": "have not",
"havent": "have not",
"he'd": "he had / he would",
"hed": "he had / he would",
"he'd've": "he would have",
"he'll": "he shall / he will",
"he'll've": "he shall have / he will have",
"he's": "he has / he is",
"how'd": "how did",
"how'd'y": "how do you",
"how'll": "how will",
"how's": "how has / how is / how does",
"I'd": "I had / I would",
"I'd've": "I would have",
"I'll": "I shall / I will",
"I'll've": "I shall have / I will have",
"I'm": "I am",
"Im": "I am",
"I've": "I have",
"Ive": "I have",
"isn't": "is not",
"isnt": "is not",
"it'd": "it had / it would",
"it'd've": "it would have",
"it'll": "it shall / it will",
"it'll've": "it shall have / it will have",
"it's": "it has / it is",
"let's": "let us",
"ma'am": "madam",
"mayn't": "may not",
"might've": "might have",
"mightn't": "might not",
"mightn't've": "might not have",
"must've": "must have",
"mustn't": "must not",
"mustn't've": "must not have",
"needn't": "need not",
"needn't've": "need not have",
"o'clock": "of the clock",
"oughtn't": "ought not",
"oughtn't've": "ought not have",
"shan't": "shall not",
"sha'n't": "shall not",
"shan't've": "shall not have",
"she'd": "she had / she would",
"she'd've": "she would have",
"she'll": "she shall / she will",
"she'll've": "she shall have / she will have",
"she's": "she has / she is",
"should've": "should have",
"shouldn't": "should not",
"shouldn't've": "should not have",
"so've": "so have",
"so's": "so as / so is",
"that'd": "that would / that had",
"that'd've": "that would have",
"that's": "that has / that is",
"there'd": "there had / there would",
"there'd've": "there would have",
"there's": "there has / there is",
"they'd": "they had / they would",
"they'd've": "they would have",
"they'll": "they shall / they will",
"they'll've": "they shall have / they will have",
"they're": "they are",
"they've": "they have",
"to've": "to have",
"wasn't": "was not",
"we'd": "we had / we would",
"we'd've": "we would have",
"we'll": "we will",
"we'll've": "we will have",
"we're": "we are",
"we've": "we have",
"weren't": "were not",
"what'll": "what shall / what will",
"what'll've": "what shall have / what will have",
"what're": "what are",
"what's": "what has / what is",
"whats": "what has / what is",
"what've": "what have",
"when's": "when has / when is",
"when've": "when have",
"where'd": "where did",
"where's": "where has / where is",
"wheres": "where has / where is",
"where've": "where have",
"who'll": "who shall / who will",
"who'll've": "who shall have / who will have",
"who's": "who has / who is",
"whos": "who has / who is",
"who've": "who have",
"why's": "why has / why is",
"whys": "why has / why is",
"why've": "why have",
"will've": "will have",
"won't": "will not",
"wont": "will not",
"won't've": "will not have",
"would've": "would have",
"wouldn't": "would not",
"wouldnt": "would not",
"wouldn't've": "would not have",
"y'all": "you all",
"y'all'd": "you all would",
"y'all'd've": "you all would have",
"y'all're": "you all are",
"y'all've": "you all have",
"you'd": "you had / you would",
"you'd've": "you would have",
"you'll": "you shall / you will",
"you'll've": "you shall have / you will have",
"you're": "you are",
"you've": "you have"
}

In [0]:
df1

Short description  ... Assignment group
0                                        [login, issue]  ...            GRP_0
1                                             [outlook]  ...            GRP_0
2                                      [cant, log, vpn]  ...            GRP_0
3                       [unable, access, hr_tool, page]  ...            GRP_0
4                                        [skype, error]  ...            GRP_0
...                                                 ...  ...              ...
8495                    [emails, not, coming, zz, mail]  ...           GRP_29
8496                        [telephony_software, issue]  ...            GRP_0
8497  [vip2, windows, password, reset, tifpdchb, ped...  ...            GRP_0
8498                [machine, nÃ, o, estÃ, funcionando]  ...           GRP_62
8499  [mehreren, pc, s, lassen, sich, verschiedene, ...  ...           GRP_49

[8496 rows x 4 columns]

In [0]:
df1['complete_desc'] = df1['Short description'] + df1['Description'] + df1['Caller']

In [0]:
df1

Short description  ...                                      complete_desc
0                                        [login, issue]  ...  [login, issue, verified, user, details, employ...
1                                             [outlook]  ...  [outlook, received, from, hmjdrvpb, komuaywn, ...
2                                      [cant, log, vpn]  ...  [cant, log, vpn, received, from, eylqgodm, ybq...
3                       [unable, access, hr_tool, page]  ...  [unable, access, hr_tool, page, unable, access...
4                                        [skype, error]  ...   [skype, error, skype, error, owlgqjme, qhcozdfx]
...                                                 ...  ...                                                ...
8495                    [emails, not, coming, zz, mail]  ...  [emails, not, coming, zz, mail, received, from...
8496                        [telephony_software, issue]  ...  [telephony_software, issue, telephony_software...
8497  [vip2, windows, password, reset, tifpdchb, ped...  ...  [vip2, windows, password, reset, tifpdchb, ped...
8498                [machine, nÃ, o, estÃ, funcionando]  ...  [machine, nÃ, o, estÃ, funcionando, unable, ac...
8499  [mehreren, pc, s, lassen, sich, verschiedene, ...  ...  [mehreren, pc, s, lassen, sich, verschiedene, ...

[8496 rows x 5 columns]

In [0]:
df1['Description'].iloc[10]

['engineering',
 'tool',
 'says',
 'not',
 'connected',
 'unable',
 'submit',
 'reports']

KeyError: ignored

In [0]:
|pip install WordCloud

In [0]:
from wordcloud import WordCloud, STOPWORDS 
    
comment_words = '' 
  
# iterate through the csv file 
for val in df1['complete_desc']: 
      
    # typecaste each val to string 
    val = str(val) 
  
    # split the value 
    tokens = val.split('\n') 
     
      
    # Converts each token into lowercase 
    for i in range(len(tokens)): 
        tokens[i] = tokens[i].lower() 
        nltk.RegexpTokenizer(r"'<[^>]*>\w+").tokenize(tokens[i])
      
    comment_words += " ".join(tokens)+" "
wordcloud = WordCloud(width = 800, height = 800, 
                background_color ='black', 
                stopwords = stop_words, 
                min_font_size = 10).generate(comment_words) 
  
# plot the WordCloud image                        
plt.figure(figsize = (8, 8), facecolor = None) 
plt.imshow(wordcloud) 
plt.axis("off") 
plt.tight_layout(pad = 0) 
  
plt.show() 

In [0]:
tokens

In [0]:
|  comment_words

In [0]:
from wordcloud import WordCloud, STOPWORDS 
    
comment_words_s = '' 
  
# iterate through the csv file 
for val in df1['Short description']: 
      
    # typecaste each val to string 
    val = str(val) 
  
    # split the value 
    tokens = val.split() 
      
    # Converts each token into lowercase 
    for i in range(len(tokens)): 
        tokens[i] = tokens[i].lower() 
        nltk.RegexpTokenizer(r"\w+").tokenize(tokens[i])
      
    comment_words_s += " ".join(tokens)+" "
wordcloud = WordCloud(width = 800, height = 800, 
                background_color ='white', 
                stopwords = stop_words, 
                min_font_size = 10).generate(comment_words_s) 
  
# plot the WordCloud image                        
plt.figure(figsize = (8, 8), facecolor = None) 
plt.imshow(wordcloud) 
plt.axis("off") 
plt.tight_layout(pad = 0) 
  
plt.show() 

In [0]:
  from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder

In [0]:
grp = df1['Assignment group']

grp

0        GRP_0
1        GRP_0
2        GRP_0
3        GRP_0
4        GRP_0
         ...  
8495    GRP_29
8496     GRP_0
8497     GRP_0
8498    GRP_62
8499    GRP_49
Name: Assignment group, Length: 8496, dtype: object

In [0]:
#df1.drop('Assignment group',1,inplace = True)

In [0]:
label = LabelEncoder()
y = label.fit_transform(df1['Assignment group'])

In [0]:
X =  df1['complete_desc']

In [0]:
#df['Assignment group'] = df['Assignment group'].str.replace("GRP_","")

In [0]:
y_encoded[478]

array([0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0.])

In [0]:
onehot_encoder = OneHotEncoder(sparse=False)
y = y.reshape(len(y), 1)
y_encoded = onehot_encoder.fit_transform(y)

In [0]:
y_encoded.shape

(8496, 74)

In [0]:

fig_dims = (25, 8)
fig, ax = plt.subplots(figsize=fig_dims)
sns.countplot(data = grp, x='Assignment group')


In [0]:
max_features = 20000
max_len = 500


In [0]:
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from sklearn.model_selection import train_test_split
#Let us try some models
from sklearn import model_selection, naive_bayes, svm
from sklearn.metrics import accuracy_score
from sklearn.metrics import accuracy_score,confusion_matrix
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB

In [0]:
tokenizer = Tokenizer(num_words=max_features)

tokenizer.fit_on_texts(df1['complete_desc'])

seq = tokenizer.texts_to_sequences(df1['complete_desc'])

In [0]:
X = pad_sequences(seq, maxlen = max_len)

In [0]:
print(tokenizer.word_index)



{'com': 1, 'not': 2, 'company': 3, 'from': 4, 'received': 5, '00': 6, '2016': 7, 'gmail': 8, 'please': 9, 'password': 10, 'erp': 11, 'job': 12, 'job_scheduler': 13, 'no': 14, 'yes': 15, 'failed': 16, 'na': 17, 'at': 18, '10': 19, 'reset': 20, 'unable': 21, 'user': 22, 'access': 23, 'issue': 24, 'account': 25, '1': 26, '09': 27, 'ticket': 28, 'error': 29, 'monitoring_tool': 30, '08': 31, 'outlook': 32, 'email': 33, 'working': 34, 'login': 35, 'site': 36, 'bpctwhsn': 37, 'kzqsbmtp': 38, 'help': 39, '2': 40, 'system': 41, 'need': 42, 'name': 43, 'sid_34': 44, 'id': 45, '16': 46, 'â': 47, 'hello': 48, 'locked': 49, '4': 50, 'update': 51, '23': 52, 'circuit': 53, 'å': 54, 'power': 55, 'new': 56, 'vpn': 57, 'windows': 58, 'network': 59, 'server': 60, '0': 61, 'able': 62, 'backup': 63, '11': 64, 's': 65, 'usa': 66, 'skype': 67, 'message': 68, 'phone': 69, 'cid': 70, 'crm': 71, 'see': 72, 'outage': 73, 'event': 74, '3': 75, '13': 76, 'log': 77, 'printer': 78, '12': 79, 'type': 80, 'request': 8

In [0]:

#Train_test split

In [0]:
X_train,X_test,y_train,y_test = train_test_split(X,y_encoded,test_size=0.3,random_state=1)
  
  

In [0]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(5947, 500)
(2549, 500)
(5947, 74)
(2549, 74)


In [0]:
#Let us try some models
from sklearn import model_selection, naive_bayes, svm
from sklearn.metrics import accuracy_score
from sklearn.metrics import accuracy_score,confusion_matrix
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.pipeline import Pipeline

In [0]:
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [0]:
print(knn.score(X_train,y_train))
print(knn.score(X_test,y_test))

0.5031108121742055
0.44801883091408395


In [0]:
Naive = GaussianNB()
Naive.fit(X_train,y_train)

ValueError: ignored

In [0]:
SVM = svm.SVC(C=1.0, kernel='linear', degree=3, gamma='auto')
SVM.fit(X_train,y_train)

ValueError: ignored

In [0]:
#def identify_tokens(row):
   #review = row['review']
    #tokens = nltk.word_tokenize(review)
    # taken only words (not punctuation)
    #token_words = [w for w in tokens if w.isalpha()]
    #return token_words

In [0]:
randomf = RandomForestClassifier(n_estimators= 150)
randomf.fit(X_train,y_train)
print("rfTree on train data ", randomf.score(X_train,y_train))
print("rfTree on train data ", randomf.score(X_test,y_test))

rfTree on train data  0.9979821758870019
rfTree on train data  0.4315417810906238


In [0]:
#Hyperparameter tuning

In [0]:
randomf1 = RandomForestClassifier(criterion = 'entropy' ,n_estimators= 150,oob_score = True,n_jobs = -1 ,random_state = 50,max_features ='auto',min_samples_leaf = 50)

In [0]:
randomf1.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='entropy', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=50, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=150,
                       n_jobs=-1, oob_score=True, random_state=50, verbose=0,
                       warm_start=False)

In [0]:
print("rfTree on train data ", randomf1.score(X_train,y_train))
print("rfTree on train data ", randomf1.score(X_test,y_test))

rfTree on train data  0.33058685051286363
rfTree on train data  0.32326402510788543
